nomad/resource_variable.go

```diff
@@ -0,0 +1,185 @@
+	d.SetId("")
+
+	return nil
+}
+
+func resourceVariableRead(d *schema.ResourceData, meta any) error {
+	client := meta.(ProviderConfig).client
+
+	path := d.Get("path").(string)
+	ns := d.Get("namespace").(string)
+	variableID := path + "@" + ns
+
+	log.Printf("[DEBUG] Reading variable %s", variableID)
+	variable, _, err := client.Variables().Read(path, &api.QueryOptions{Namespace: ns})
+	if err != nil {
+		return fmt.Errorf("error getting information about %s: %v", variableID, err)
+	}
+
+	d.SetId(variableID)
+	return d.Set("items", variable.Items)
+}
+
+func resourceVariableExists(d *schema.ResourceData, meta any) (bool, error) {
+	client := meta.(ProviderConfig).client
+	variableID := d.Id()
+
+	log.Printf("[DEBUG] Checking if Variable %q exists", variableID)
+
+	path := d.Get("path").(string)
+	ns := d.Get("namespace").(string)
+
+	_, _, err := client.Variables().Read(path, &api.QueryOptions{Namespace: ns})
+	if err != nil {
+		// As of Nomad 0.4.1, the API client returns an error for 404
+		// rather than a nil result, so we must check this way.
+		if strings.Contains(err.Error(), "404") {
+			return false, nil
+		}
+
+		return true, fmt.Errorf("error checking for Variable %s: %#v", variableID, err)
+	}
+
+	return true, nil
+}
+
+func pathValidation() schema.SchemaValidateFunc {
+	return func(i interface{}, k string) ([]string, []error) {
+		// Verify path actually is a string
+		path, ok := i.(string)
+		if !ok {
+			return nil, []error{fmt.Errorf("expected type of %s to be string", k)}
+		}
+
+		// Limit length to 128 characters
+		if len(path) > maxPathLength {
+			return nil, []error{fmt.Errorf("expecnted path legnth to be less than %v but got a path legnth of %v", maxPathLength, len(path))}
+		}
+
+		// Limit to RFC3986 URL-safe characters, minus '@' and '.' as they conflict with template blocks
+		if !validVariablePath.MatchString(path) {
+			return nil, []error{fmt.Errorf("the path %s contains invalid characters", path)}
+		}
+
+		// Validate paths that have 'nomad' as the first directory in the path
+		// The following validation is taken from:
+		// https://github.com/hashicorp/nomad/blob/c7c8bd4ce2fcd9f2cd13278043b7baf1f8dd4f47/nomad/structs/variables.go#L191
+		parts := strings.Split(path, "/")
+
+		if parts[0] != "nomad" {
+			return nil, nil
+		}
+
+		if len(parts) == 1 {
+			return nil, []error{fmt.Errorf("path of 'nomad' is a reserved top-level directory path")}
+		}
+
+		switch {
+		case parts[1] == "jobs":
+			return nil, nil
+		case parts[1] == "job-templates" && len(parts) == 3:
+			return nil, nil
+		case parts[1] == "job-templates":
+			return nil, []error{fmt.Errorf("the path 'nomad/job-templates' is reserved, you may write variables at the level below it, for example, 'nomad/job-templates/template-name'")}
+		default:
+			return nil, []error{fmt.Errorf("only paths at 'nomad/jobs' or 'nomad/job-templates' and below are valid paths under the top-level 'nomad' directory")}
+		}
+	}
+}
```

Comment:
Since Terrafor validation allows returning multiple errors, it's always best to check for and return as many errors as possible. This avoids the bad UX of users having to go through multiple cycles of `apply-error-fix-apply-error-fix-...`.

In this case I think only the first error is required to return early, so the function could look something like this:

```go
func pathValidation() schema.SchemaValidateFunc {
	return func(i interface{}, k string) ([]string, []error) {
		// Verify path actually is a string
		path, ok := i.(string)
		if !ok {
			return nil, []error{fmt.Errorf("expected type of %s to be string", k)}
		}

		errs := []error{}

		// Limit length to 128 characters
		if len(path) > maxPathLength {
			errs = append(errs, fmt.Errorf("expecnted path legnth to be less than %v but got a path legnth of %v", maxPathLength, len(path)))
		}

		// Limit to RFC3986 URL-safe characters, minus '@' and '.' as they conflict with template blocks
		if !validVariablePath.MatchString(path) {
			errs = append(errs, fmt.Errorf("the path %s contains invalid characters", path))
		}

		// Validate paths that have 'nomad' as the first directory in the path
		// The following validation is taken from:
		// https://github.com/hashicorp/nomad/blob/c7c8bd4ce2fcd9f2cd13278043b7baf1f8dd4f47/nomad/structs/variables.go#L191
		parts := strings.Split(path, "/")

		if parts[0] != "nomad" {
			return nil, errs
		}

		if len(parts) == 1 {
			errs = append(errs, fmt.Errorf("path of 'nomad' is a reserved top-level directory path"))
		}

		switch {
		case parts[1] == "jobs":
			return nil, errs
		case parts[1] == "job-templates" && len(parts) == 3:
			return nil, errs
		case parts[1] == "job-templates":
			errs = append(errs, fmt.Errorf("the path 'nomad/job-templates' is reserved, you may write variables at the level below it, for example, 'nomad/job-templates/template-name'"))
		default:
			errs = append(errs, fmt.Errorf("only paths at 'nomad/jobs' or 'nomad/job-templates' and below are valid paths under the top-level 'nomad' directory"))
		}
		return nil, errs
	}
}
```

Reply:
I did not realize I could return multiple errors like that. Thank you very much for pointing that out. It would indeed make it might nicer to get a list of errors at once.